query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/converters/PinotFunctionConverter.java

```diff
@@ -33,25 +39,67 @@ public PinotFunctionConverter() {
   }
 
-  public String convert(
-      Function function, java.util.function.Function<Expression, String> argumentConverter) {
+  public String convert(ExecutionContext executionContext, Function function, java.util.function.Function<Expression, String> argumentConverter) {
     switch (function.getFunctionName().toUpperCase()) {
       case QUERY_FUNCTION_COUNT:
         return this.convertCount();
       case QUERY_FUNCTION_PERCENTILE:
         return this.functionToString(this.toPinotPercentile(function), argumentConverter);
       case QUERY_FUNCTION_CONCAT:
         return this.functionToString(this.toPinotConcat(function), argumentConverter);
+      case QUERY_FUNCTION_AVG_RATE:
+        return this.functionToStringForAvgRate(function, argumentConverter, executionContext);
       default:
         // TODO remove once pinot-specific logic removed from gateway - this normalization reverts
         // that logic
         if (this.isHardcodedPercentile(function)) {
-          return this.convert(this.normalizeHardcodedPercentile(function), argumentConverter);
+          return this.convert(executionContext, this.normalizeHardcodedPercentile(function), argumentConverter);
         }
         return this.functionToString(function, argumentConverter);
     }
   }
 
+  private String functionToStringForAvgRate(Function function, java.util.function.Function<Expression, String> argumentConverter, ExecutionContext executionContext) {
+    function = updateFunctionForAvgRate(function, executionContext);
+    String argumentString =
+        function.getArgumentsList().stream()
+            .map(argumentConverter::apply)
+            .collect(Collectors.joining(","));
+
+    return "SUM(DIV(" + argumentString + "))";
+  }
+
+  private Function updateFunctionForAvgRate(Function function, ExecutionContext executionContext){
+
+    Expression columnName = function.getArgumentsList().get(0);
+    Expression literal = function.getArgumentsList().get(1);
+
+    Builder builder = function.toBuilder();
+    builder.clearArguments();
+    builder.addArguments(columnName);
+    builder.addArguments(getUpdatedLiteral(columnName, literal, executionContext));
+
+    return builder.build();
+  }
+
+  private Expression getUpdatedLiteral(Expression column, Expression literal, ExecutionContext executionContext){
+
+    String columnName = column.getColumnIdentifier().getColumnName().split("[.]")[0];
+    long divisorInSeconds = literal.getLiteral().getValue().getLong();
+    double timeRangeInSeconds;
+
+    if(executionContext.getTimeSeriesColumnMap().containsKey(columnName)){
+      timeRangeInSeconds = Double.parseDouble(executionContext.getTimeSeriesColumnMap().get(columnName));
+    }
+    else{
+      timeRangeInSeconds = (double) executionContext.getTimeFilterMap().get(columnName).get(1)-executionContext.getTimeFilterMap().get(columnName).get(0);
+      timeRangeInSeconds = timeRangeInSeconds / 1000 ;
```

Comment:
So this looks like we're doing both of:
- assigning a value to "timeRangeInSeconds" that isn't actually in seconds. Variables are cheap, we shouldn't reuse an existing one if the name doesn't make sense for the value we're putting in it.
- Assuming that the time range is in millis, also not an assumption we want to make if we can help it (and if we need to, we should isolate it and document)